crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+        match self.operator {
+            //convert greater into less
+            Operation::ugt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ult
+            }
+            Operation::uge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ule
+            }
+            Operation::sgt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::slt
+            }
+            Operation::sge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::sle
+            }
+            Operation::gt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lt
+            }
+            Operation::gte => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lte
+            }
+            //TODO replace a<b with a<=b+1, but beware of edge cases!
+            Operation::eq_gate => {
+                if self.rhs == self.lhs {
+                    self.rhs = self.idx;
+                    self.is_deleted = true;
+                }
+            }
+            _ => (),
+        }
+        if is_commutative(self.operator) && self.rhs < self.lhs {
+            std::mem::swap(&mut self.rhs, &mut self.lhs);
+        }
+    }
+}
+
+//adpated from LLVM IR
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Operation {
+    add,      //(+)
+    safe_add, //(+) safe addtion
+    sub,      //(-)
+    safe_sub, //(-) safe substraction
+    mul,      //(*)
+    safe_mul, //(*) safe multiplication
+    udiv,     //(/) unsigned division
+    sdiv,     //(/) signed division
+    urem,     //(%) modulo; remainder of unsigned division
+    srem,     //(%) remainder of signed division
+    fmod,     //(%) remainder of the floating point division
+    fneg,     //(-) negation of a float
+    fdiv,     //(/) floating point division
+    div,      //(/) field division
+    eq,       //(==) equal
+    ne,       //(!=) not equal
+    ugt,      //(>) unsigned greater than
+    uge,      //(>=) unsigned greater or equal
+    ult,      //(<) unsigned less than
+    ule,      //(<=) unsigned less or equal
+    sgt,      //(>) signed greater than
+    sge,      //(>=) signed greater or equal
+    slt,      //(<) signed less than
+    sle,      //(<=) signed less or equal
+    lt,       //(<) field less
+    gt,       //(>) field greater
+    lte,      //(<=) field less or equal
+    gte,      //(<=) field greater or equal
+    and,
+    not,
+    or,
+    xor,
+    cast,  //convert type
+    ass,   //assignement
+    trunc, //truncate
+
+    //control flow
+    jne, //jump on not equal
+    jeq, //jump on equal
+    jmp, //unconditional jump
+    phi,
+    // todo: call, br,..
```

Comment:
lets remove this todo and open up an issue for it

Reply:
There are times where we pass in a mutable reference, and the reason to do this is to save on heap allocations for large buffers. For example: https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string

Even with that read_to_string, we have this method which wraps around it: https://doc.rust-lang.org/std/fs/fn.read_to_string.html . (Not to say people use the latter over the former)

I'd also not use this way, if the function is called once on a buffer because then you are not saving much. In this particular case, I don't think we are saving anything by passing in a `&mut Integer` and `&mut bool`. 

Note that we also do not do this in the rest of the code unless its for ownership reasons, its never because we want to optimise heap allocations, which seems to be a more common approach in C++ for example. (Maybe there are some places where we do need to do it due to large buffer re-allocations)